source/adapters/level_zero/command_buffer.cpp

```diff
@@ -614,9 +799,29 @@ urCommandBufferFinalizeExp(ur_exp_command_buffer_handle_t CommandBuffer) {
              (CommandBuffer->ZeCommandListResetEvents,
               CommandBuffer->AllResetEvent->ZeEvent));
 
+  return UR_RESULT_SUCCESS;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL
+urCommandBufferFinalizeExp(ur_exp_command_buffer_handle_t CommandBuffer) {
+  UR_ASSERT(CommandBuffer, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+
+  // It is not allowed to append to command list from multiple threads.
+  std::scoped_lock<ur_shared_mutex> Guard(CommandBuffer->Mutex);
+
+  if (checkImmediateAppendSupport(CommandBuffer->Context,
```

Comment:
We already queried this when we created the command buffer, does it make sense to just store that in the command buffer itself or infer from existing info in the command buffer? SInce the device and context cannot change here.